### PR TITLE
Update Commit config to separate add/version and add back skipCi

### PR DESCRIPTION
**Pull request by rohit-gohri**
PoC for a possible solution for #688. Made commit option configurable separately for `add`/`version` commands. Adds back `skipCI` option

```
{
  "commit": true
}
// above evaluates to below
{
  "commit": {
    "add": {
      "skipCI": false
    },
    "version": {
      "skipCI": true
    }
  }
}
```

Other ways:
```
// only add
{
  "commit": {
    "add": {
      "skipCI": false
    },
    "version":false
  }
}
// only version
{
  "commit": {
    "add": false,
    "version": {
      "skipCI": false
    }
  }
}
```

### CODE REVIEW

**Comment by Andarist** [comment]
Thank you for the PR - I'll try to do a review as soon as possible but I have lots on my plate right now and can't get to everything right away (even though I would really like to 😛 )

**Comment by Andarist** [comment]
I see how this solution is more robust but I'm not yet sure if we want to allow people to hook into such stuff. I think that **for now** I would prefer if we'd go with a simpler, builtin, solution.

I imagine that we could support this:
```json
{
  "commit": {
    "version": {
      "skipCI": true
    },
    "publish": {
      "skipCI": true
    }
  }
}
```

We'd have to make sure that we use the same default values as the current, implicit, ones. A config like this should only enable auto-commits for the version command:
```json
{
  "commit": {
    "version": true
  }
}
```

**Comment by rohit-gohri** [comment]
@Andarist Have updated the PR to an approach similar to what you described

**Comment by Andarist** [comment]
Thank you for the PR - I will try to review it after the weekend

File: packages/types/src/index.ts [hunk 1]
```diff
@@ -58,9 +58,18 @@ export type PackageJSON = {
...
+
+export type Commit = {
+  add: false | CommitOptions;
+  version: false | CommitOptions;
+};
+
 export type Config = {
   changelog: false | readonly [string, any];
-  commit: boolean;
+  commit: Commit;
```
Comment (Andarist) [inline packages/types/src/index.ts:72]: This is a breaking change for all packages that receive config from the outside, such as `@changesets/apply-release-plan`. Previously they could assume that this is a boolean. I think it's not worth a breaking change and perhaps we don't have to normalize this (at least not for the time being) to this object shape.

File: packages/config/src/index.ts [hunk 2]
```diff
@@ -359,3 +397,5 @@ export let defaultConfig = parse(defaultWrittenConfig, {
   tool: "root",
   packages: [fakePackage]
 });
+
+export { getAddLine, getVersionLine };
```
Comment (Andarist) [inline packages/config/src/index.ts:401]: I think that those don't belong in this package, at least for now - moving those should be reverted, they should stay declared closer to where they are actually called.

File: packages/config/src/index.ts [hunk 3]
```diff
@@ -113,14 +118,45 @@ export let parse = (json: WrittenConfig, packages: Packages): Config => {
...
+  ) {
     messages.push(
       `The \`commit\` option is set as ${JSON.stringify(
         json.commit,
         null,
         2
-      )} when the only valid values are undefined or a boolean`
+      )} when the only valid values are undefined or a boolean or an object`
     );
+  } else if (typeof json.commit === "object") {
```
Comment (Andarist) [inline packages/config/src/index.ts:133]: we could also add a check if `skipCI` is of the correct type within this block

**Comment by Andarist** [comment]
Actually… while taking a shower i was thinking about this more 😅and i think i’ve figured out an alternative, more robust, design. Gonna post it later today so maybe let’s hold working on this PR until i write it down

**Comment by Andarist** [comment]
My proposal is to accept a string for the commit option that would point to a module responsible for generating those commit messages. We already do a similar thing with "changelog generators":
https://github.com/changesets/changesets/blob/c2b9d44886740f850704b06a4d15beca2ca3a1d8/packages/apply-release-plan/src/index.ts#L201-L223

I feel like this would be the most robust option that wouldn't require us to add a somewhat weird shape of the `commit` config option. This way the default module for this (that probably should be just an additional entry in the `@changesets/cli`) could not add `[skip ci]` at all but it would be super easy to create a module that would call the default one and just adjust the return value slightly. I do a similar thing with the default changelog generator here:
https://github.com/contentlayerdev/contentlayer/blob/7c06f352e73e265ad57e0ef12c88b892bc85837d/scripts/changeset-changelog-generator.js

Note that when using the default "commit generator" we'd have to add `[skip ci]` to the message associated with one of the commands to maintain backward compatibility. This could happen in the code and outside of this default module though - unless this would be too cumbersome to pull off in the codebase, then it's just fine to include that in the default module.

Open questions:
- what arguments function(s) exposed by this module should receive?
- should the module expose a single function or two? it's important to know the content of which command we are generating but this information could be given in an argument, so using a single function is an option. I didn't think through which one should be preferred here though.

**Comment by rohit-gohri** [comment]
> My proposal is to accept a string for the commit option that would point to a module responsible for generating those commit messages.
Ha, that was the orignal PR's approach. It's still in this old branch: https://github.com/changesets/changesets/compare/main...rohit-gohri:old
I should have just opened a new PR 🤦  but I force-pushed instead to keep the discussion in single place.

Though I didn't move it into it's own module, I'll open a new PR with the changes.

> what arguments function(s) exposed by this module should receive?
the changesets for the `add` and the release info for the `publish`, and any config?

> should the module expose a single function or two?
I think 2 makes more sense as the input arguments would be completely different.

**Comment by Andarist** [comment]
> Ha, that was the orignal PR's approach. It's still in this old branch: https://github.com/changesets/changesets/compare/main...rohit-gohri:old

Heh, ye - I've reconsidered my previous opinion after getting a better look at what the alternative actually looks like 😅 

> Though I didn't move it into it's own module

It doesn't have to be a new module - an additional exntrypoint in the `@changesets/cli` should be sufficient. Take a look at how the one for the changelog is being defined here:
https://github.com/changesets/changesets/blob/0d1debd8f5d364073d627a017316a25554dec4d5/packages/cli/package.json#L26

> I'll open a new PR with the changes.

Feel free to force-push to this one if you prefer that.

> the changesets for the add and the release info for the publish, and any config?

Yep, sounds alright.

> I think 2 makes more sense as the input arguments would be completely different.

Agreed. Such functions could actually be typed somewhat OK with the upcoming [dependent parameters](https://devblogs.microsoft.com/typescript/announcing-typescript-4-6-rc/#dependent-parameters-cfa) but ye, it's probably more straightforward to use separate functions.

**Comment by rohit-gohri** [comment]
Opened https://github.com/changesets/changesets/pull/768 based on the discussion here, didn't force push so that we can compare the approaches. Closing this.

**Pull request closed without merging**